query: v0.29.0. Opentracing config create gopanic SIGSEGV

**Thanos, Prometheus and Golang version used**:
thanos v0.29.0

**What happened**:
While upgrading thanos version from 0.28.1 to v0.29.0, all our thanos queriers went in Crashloop because of a Segmentation violation (see logs below)

**What you expected to happen**:
as per https://github.com/thanos-io/thanos/pull/5411 states, there should be no breaking change.
Removing the tracing config from the specs, fix the pod (but remove tracing :) )
```yaml
            - |
              --tracing.config=type: JAEGER
              config:
                service_name: "thanos-query-main"
                disabled: false
                rpc_metrics: false
                tags: ""
                sampler_type: ""
                sampler_param: 1
                sampler_manager_host_port: ""
                sampler_max_operations: 0
                sampler_refresh_interval: 0s
                reporter_max_queue_size: 0
                reporter_flush_interval: 0s
                reporter_log_spans: false
                endpoint: ""
                user: ""
                password: ""
                agent_host: "jaeger-main-agent.argos.svc"
                agent_port: 6831
```

**How to reproduce it (as minimally and precisely as possible)**:
hard to tell, on our setup with the working opentracing config above, we just upgraded the docker image version... :/
*I am not sure about the healthiness of my jaeger installation though*, maybe it panics when the otel target is down only. Still this should not break the thanos process if this happens :)

**Full logs to relevant components**:

<details>Logs
<p>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x143d0e8]

goroutine 456 [running]:
go.opentelemetry.io/otel/sdk/trace.parentBased.ShouldSample({{0x0, 0x0}, {{0x270e3f0, 0x3a18158}, {0x270e3c8, 0x3a18158}, {0x270e3f0, 0x3a18158}, {0x270e3c8, 0x3a18158}}}, ...)
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.10.0/trace/sampling.go:281 +0x1c8
github.com/thanos-io/thanos/pkg/tracing/migration.samplerWithOverride.ShouldSample(...)
	/app/pkg/tracing/migration/sampler.go:42
go.opentelemetry.io/otel/sdk/trace.(*tracer).newSpan(0xc0005b0640, {0x2719670, 0xc000e97980}, {0xc0001a0d50, 0x16}, 0xc000ec3480)
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.10.0/trace/tracer.go:95 +0x456
go.opentelemetry.io/otel/sdk/trace.(*tracer).Start(0xc0005b0640, {0x2719670?, 0xc000e97980?}, {0xc0001a0d50, 0x16}, {0xc000e9af40?, 0x4?, 0xc000ec35b0?})
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.10.0/trace/tracer.go:52 +0x153
go.opentelemetry.io/otel/bridge/opentracing.(*WrapperTracer).Start(0xc0006b6228, {0x2719670?, 0xc000e97980?}, {0xc0001a0d50?, 0x1eae540?}, {0xc000e9af40?, 0x1?, 0xc000ea2f70?})
	/go/pkg/mod/go.opentelemetry.io/otel/bridge/opentracing@v1.10.0/wrapper.go:79 +0x4b
go.opentelemetry.io/otel/bridge/opentracing.(*BridgeTracer).StartSpan(0xc0006ce3c0, {0xc0001a0d50, 0x16}, {0xc000e97920, 0x3, 0x3a18158?})
	/go/pkg/mod/go.opentelemetry.io/otel/bridge/opentracing@v1.10.0/bridge.go:430 +0x3f5
github.com/thanos-io/thanos/pkg/tracing/migration.(*bridgeTracerWrapper).StartSpan(0x2009940?, {0xc0001a0d50?, 0x3a18158?}, {0xc000e97920?, 0x39acd40?, 0xc000ec3828?})
	/app/pkg/tracing/migration/bridge.go:89 +0x26
github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/tracing.newClientSpanFromContext({0x2719638, 0xc000ed88a0}, {0x2714460, 0xc00030a080}, {0xc0001a0d50, 0x16})
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/interceptors/tracing/client.go:92 +0x245
github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/tracing.(*opentracingClientReportable).ClientReporter(0xc0006b64e0, {0x2719638, 0xc000ed88a0}, {0x0?, 0x0?}, {0x220b8e2, 0x5}, {0x22309c5, 0x10}, {0x22309d6, ...})
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/interceptors/tracing/client.go:51 +0x127
github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors.UnaryClientInterceptor.func1({0x2719638, 0xc000ed88a0}, {0x22309c4, 0x16}, {0x21122a0, 0x3a18158}, {0x21123e0, 0xc000ed89c0}, 0xc000ea0c30?, 0x22f4818, ...)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/interceptors/client.go:19 +0x195
github.com/grpc-ecosystem/go-grpc-middleware/v2.ChainUnaryClient.func1.1.1({0x2719638?, 0xc000ed88a0?}, {0x22309c4?, 0x38?}, {0x21122a0?, 0x3a18158?}, {0x21123e0?, 0xc000ed89c0?}, 0x0?, {0xc000eb47c0, ...})
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/chain.go:74 +0x86
github.com/grpc-ecosystem/go-grpc-prometheus.(*ClientMetrics).UnaryClientInterceptor.func1({0x2719638, 0xc000ed88a0}, {0x22309c4, 0x16}, {0x21122a0, 0x3a18158}, {0x21123e0, 0xc000ed89c0}, 0x8?, 0xc000e98a68, ...)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/client_metrics.go:112 +0x117
github.com/grpc-ecosystem/go-grpc-middleware/v2.ChainUnaryClient.func1.1.1({0x2719638?, 0xc000ed88a0?}, {0x22309c4?, 0x203000?}, {0x21122a0?, 0x3a18158?}, {0x21123e0?, 0xc000ed89c0?}, 0x1?, {0xc000eb47c0, ...})
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/chain.go:74 +0x86
github.com/grpc-ecosystem/go-grpc-middleware/v2.ChainUnaryClient.func1({0x2719638, 0xc000ed88a0}, {0x22309c4, 0x16}, {0x21122a0, 0x3a18158}, {0x21123e0, 0xc000ed89c0}, 0x0?, 0x22f4818, ...)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/chain.go:83 +0x157
google.golang.org/grpc.(*ClientConn).Invoke(0xc00085ea00?, {0x2719638?, 0xc000ed88a0?}, {0x22309c4?, 0x16?}, {0x21122a0?, 0x3a18158?}, {0x21123e0?, 0xc000ed89c0?}, {0xc000e94a10, ...})
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/call.go:35 +0x223
github.com/thanos-io/thanos/pkg/info/infopb.(*infoClient).Info(0xc00030a2c8, {0x2719638, 0xc000ed88a0}, 0x8?, {0xc000e94a10, 0x1, 0x1})
	/app/pkg/info/infopb/rpc.pb.go:422 +0xc9
github.com/thanos-io/thanos/pkg/query.(*endpointRef).Metadata(0xc0006fca80, {0x2719638, 0xc000ed88a0}, {0x2700800, 0xc00030a2c8}, {0x271a8d0, 0xc00030a2d0})
	/app/pkg/query/endpointset.go:61 +0xe3
github.com/thanos-io/thanos/pkg/query.(*EndpointSet).updateEndpoint(0xc000a48300, {0x2719638, 0xc000ed88a0}, 0xc000e987c8, 0xc0006fca80)
	/app/pkg/query/endpointset.go:409 +0x105
github.com/thanos-io/thanos/pkg/query.(*EndpointSet).Update.func2(0xc000e987c8)
	/app/pkg/query/endpointset.go:349 +0x2cb
created by github.com/thanos-io/thanos/pkg/query.(*EndpointSet).Update
	/app/pkg/query/endpointset.go:338 +0x60a
```

</p>
</details>

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -3,6 +3,7 @@ version: 2.1
 
 orbs:
   go: circleci/go@1.7.1
+  git-shallow-clone: guitarrapc/git-shallow-clone@2.4.0
 
 executors:
   golang:
@@ -19,7 +20,7 @@ jobs:
     environment:
       GO111MODULE: 'on'
     steps:
-      - checkout
+      - git-shallow-clone/checkout
       - go/mod-download-cached
       - setup_remote_docker:
           version: 20.10.12
@@ -58,7 +59,7 @@ jobs:
       GOBIN: "/home/circleci/.go_workspace/go/bin"
       PROMU_VERSION: "0.5.0"
     steps:
-      - checkout
+      - git-shallow-clone/checkout
       - run: mkdir -p ${GOBIN}
       - run: curl -L "https://github.com/prometheus/promu/releases/download/v${PROMU_VERSION}/promu-${PROMU_VERSION}.$(go env GOOS)-$(go env GOARCH).tar.gz" | tar --strip-components=1 -xzf - -C ${GOBIN}
       - run: mv -f ${GOBIN}/promu "${GOBIN}/promu-v${PROMU_VERSION}"
@@ -71,7 +72,7 @@ jobs:
   publish_main:
     executor: golang
     steps:
-      - checkout
+      - git-shallow-clone/checkout
       - go/mod-download-cached
       - setup_remote_docker:
           version: 20.10.12
@@ -93,7 +94,7 @@ jobs:
   publish_release:
     executor: golang
     steps:
-      - checkout
+      - git-shallow-clone/checkout
       - go/mod-download-cached
       - setup_remote_docker:
           version: 20.10.12

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,18 @@ NOTE: As semantic versioning states all 0.y.z releases can contain breaking chan
 
 We use *breaking :warning:* to mark changes that are not backward compatible (relates only to v0.y.z releases.)
 
+## Unreleased
+
+## [v0.29.1](https://github.com/thanos-io/thanos/tree/release-0.29) - In progress
+
+### Fixed
+
+- [#5887](https://github.com/thanos-io/thanos/pull/5887) Tracing: Make sure rate limiting sampler is the default, as was the case in version pre-0.29.0.
+
+### Added
+
+### Changed
+
 ## [v0.29.0](https://github.com/thanos-io/thanos/tree/release-0.29) - 2022.11.03
 
 ### Fixed

--- a/pkg/tracing/jaeger/config_yaml.go
+++ b/pkg/tracing/jaeger/config_yaml.go
@@ -19,6 +19,13 @@ import (
 	tracesdk "go.opentelemetry.io/otel/sdk/trace"
 )
 
+const (
+	SamplerTypeRemote        = "remote"
+	SamplerTypeProbabilistic = "probabilistic"
+	SamplerTypeConstant      = "const"
+	SamplerTypeRateLimiting  = "ratelimiting"
+)
+
 type ParentBasedSamplerConfig struct {
 	LocalParentSampled  bool `yaml:"local_parent_sampled"`
 	RemoteParentSampled bool `yaml:"remote_parent_sampled"`
@@ -114,22 +121,27 @@ func getSamplingFraction(samplerType string, samplingFactor float64) float64 {
 
 func getSampler(config Config) tracesdk.Sampler {
 	samplerType := config.SamplerType
+	if samplerType == "" {
+		samplerType = SamplerTypeRateLimiting
+	}
 	samplingFraction := getSamplingFraction(samplerType, config.SamplerParam)
 
 	var sampler tracesdk.Sampler
 	switch samplerType {
-	case "probabilistic":
-		sampler = tracesdk.ParentBased(tracesdk.TraceIDRatioBased(samplingFraction))
-	case "const":
+	case SamplerTypeProbabilistic:
+		sampler = tracesdk.TraceIDRatioBased(samplingFraction)
+	case SamplerTypeConstant:
 		if samplingFraction == 1.0 {
 			sampler = tracesdk.AlwaysSample()
 		} else {
 			sampler = tracesdk.NeverSample()
 		}
-	case "remote":
+	case SamplerTypeRemote:
 		remoteOptions := getRemoteOptions(config)
 		sampler = jaegerremote.New(config.ServiceName, remoteOptions...)
-	case "ratelimiting":
+	// Fallback always to default (rate limiting).
+	case SamplerTypeRateLimiting:
+	default:
 		// The same config options are applicable to both remote and rate-limiting samplers.
 		remoteOptions := getRemoteOptions(config)
 		sampler = jaegerremote.New(config.ServiceName, remoteOptions...)
@@ -137,17 +149,20 @@ func getSampler(config Config) tracesdk.Sampler {
 		if ok {
 			sampler.Update(config.SamplerParam)
 		}
-	default:
-		var root tracesdk.Sampler
-		var parentOptions []tracesdk.ParentBasedSamplerOption
-		if config.SamplerParentConfig.LocalParentSampled {
-			parentOptions = append(parentOptions, tracesdk.WithLocalParentSampled(root))
-		}
-		if config.SamplerParentConfig.RemoteParentSampled {
-			parentOptions = append(parentOptions, tracesdk.WithRemoteParentSampled(root))
-		}
-		sampler = tracesdk.ParentBased(root, parentOptions...)
 	}
+
+	// Use parent-based to make sure we respect the span parent, if
+	// it is sampled. Optionally, allow user to specify the
+	// parent-based options.
+	var parentOptions []tracesdk.ParentBasedSamplerOption
+	if config.SamplerParentConfig.LocalParentSampled {
+		parentOptions = append(parentOptions, tracesdk.WithLocalParentSampled(sampler))
+	}
+	if config.SamplerParentConfig.RemoteParentSampled {
+		parentOptions = append(parentOptions, tracesdk.WithRemoteParentSampled(sampler))
+	}
+	sampler = tracesdk.ParentBased(sampler, parentOptions...)
+
 	return sampler
 }
 

--- a/pkg/tracing/jaeger/jaeger_test.go
+++ b/pkg/tracing/jaeger/jaeger_test.go
@@ -24,7 +24,7 @@ var parentConfig = ParentBasedSamplerConfig{LocalParentSampled: true}
 
 // This test shows that if sample factor will enable tracing on client process, even when it would be disabled on server
 // it will be still enabled for all spans within this span.
-func TestContextTracing_ClientEnablesTracing(t *testing.T) {
+func TestContextTracing_ClientEnablesProbabilisticTracing(t *testing.T) {
 	exp := tracetest.NewInMemoryExporter()
 	config := Config{
 		SamplerType:         "probabilistic",
@@ -65,7 +65,7 @@ func TestContextTracing_ClientEnablesTracing(t *testing.T) {
 
 // This test shows that if sample factor will disable tracing on client process,  when it would be enabled on server
 // it will be still disabled for all spans within this span.
-func TestContextTracing_ClientDisablesTracing(t *testing.T) {
+func TestContextTracing_ClientDisablesProbabilisticTracing(t *testing.T) {
 	exp := tracetest.NewInMemoryExporter()
 
 	config := Config{
@@ -87,6 +87,45 @@ func TestContextTracing_ClientDisablesTracing(t *testing.T) {
 	clientRoot, clientCtx := tracing.StartSpan(tracing.ContextWithTracer(context.Background(), tracer), "a")
 
 	config.SamplerParam = 1.0
+	sampler2 := getSampler(config)
+	// Simulate Server process with different tracer, but with client span in context.
+	srvTracerOtel := newTraceProvider(
+		context.Background(),
+		"srvTracerOtel",
+		log.NewNopLogger(),
+		tracesdk.NewSimpleSpanProcessor(exp),
+		sampler2, // never sample
+		[]attribute.KeyValue{},
+	)
+	srvTracer, _ := migration.Bridge(srvTracerOtel, log.NewNopLogger())
+
+	srvRoot, srvCtx := tracing.StartSpan(tracing.ContextWithTracer(clientCtx, srvTracer), "b")
+	srvChild, _ := tracing.StartSpan(srvCtx, "bb")
+
+	tracing.ContextTracing_ClientDisablesTracing(t, exp, clientRoot, srvRoot, srvChild)
+}
+
+func TestContextTracing_ClientDisablesAlwaysOnSampling(t *testing.T) {
+	exp := tracetest.NewInMemoryExporter()
+
+	config := Config{
+		SamplerType:  SamplerTypeConstant,
+		SamplerParam: 0,
+	}
+	sampler := getSampler(config)
+	tracerOtel := newTraceProvider(
+		context.Background(),
+		"tracerOtel",
+		log.NewNopLogger(),
+		tracesdk.NewSimpleSpanProcessor(exp),
+		sampler, // never sample
+		[]attribute.KeyValue{},
+	)
+	tracer, _ := migration.Bridge(tracerOtel, log.NewNopLogger())
+
+	clientRoot, clientCtx := tracing.StartSpan(tracing.ContextWithTracer(context.Background(), tracer), "a")
+
+	config.SamplerParam = 1
 	sampler2 := getSampler(config)
 	// Simulate Server process with different tracer, but with client span in context.
 	srvTracerOtel := newTraceProvider(

